### PR TITLE
Add types for @mysql/xdevapi

### PR DESCRIPTION
**Pull request by r-a303931**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [X] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [X] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [X] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [X] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/mysql__xdevapi/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,493 @@
+// Type definitions for @mysql/xdevapi 8.0
+// Project: https://github.com/mysql/mysql-connector-nodejs (Does not have to be to GitHub, but prefer linking to a source code repository rather than to a project website.)
```
Comment (plantain-00) [inline types/mysql__xdevapi/index.d.ts]: It should be just `// Project: https://github.com/mysql/mysql-connector-nodejs` here

**Review by plantain-00 — CHANGES_REQUESTED** [review]

**Review by plantain-00 — APPROVED** [review]

File: types/mysql__xdevapi/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+}
+
+export interface IndexDefinition {
+	type?: 'INDEX' | 'SPATIAL';
+	fields: FieldDefinition[];
+}
+
+export interface SocketFactory {
+	// tslint:disable-next-line
+	createSocket: Function;
```
Comment (uniqueiniquity) [inline types/mysql__xdevapi/index.d.ts]: As discussed [here](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes), it's not good practice to use `Function`.

File: types/mysql__xdevapi/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+
+	getGeneratedIds(): string[];
+
+	getWarnings(): Promise<Warning[]>;
+
+	getWarningsCount(): number;
+}
+
+export class Schema implements DatabaseObject {
+	// tslint:disable-next-line
```
Comment (uniqueiniquity) [inline types/mysql__xdevapi/index.d.ts]: What TSLint rule are you disabling here and why?

File: types/mysql__xdevapi/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+	createCollection<T>(
+		name: string,
+		options: CreateCollectionOptions
+	// tslint:disable-next-line
```
Comment (uniqueiniquity) [inline types/mysql__xdevapi/index.d.ts]: Same here

File: types/mysql__xdevapi/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+	lockExclusive(mode: LockContention): CollectionFind<T>;
+
+	lockShared(mode: LockContention): CollectionFind<T>;
+}
+
+export class CollectionModify<T> implements Binding, CollectionOrdering, Limiting {
+	arrayAppend<
+		K extends keyof T = keyof T,
+		// tslint:disable-next-line:no-unnecessary-generics
+		J extends keyof T[K] = keyof T[K]
```
Comment (uniqueiniquity) [inline types/mysql__xdevapi/index.d.ts]: I don't think you need these defaults here (e.g. `= keyof T`).
In fact, you don't even need `J`; you can just index with `keyof T[K]`.

File: types/mysql__xdevapi/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+	): Promise<Collection<T>>;
+
+	dropCollection(name: string): Promise<boolean>;
+
+	existsInDatabase(): Promise<boolean>;
+
+	getClassName(): string;
+
+	// tslint:disable-next-line
+	getCollection<T>(name: string): Collection<T>;
```
Comment (uniqueiniquity) [inline types/mysql__xdevapi/index.d.ts]: As discussed [here](https://github.com/Microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md), it's cleaner if this is just `getCollection(name: string): Collection<any>`.

File: types/mysql__xdevapi/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+
+	// tslint:disable-next-line
+	getCollectionAsTable<T>(name: string): Table<T>;
+
+	getCollections(): Promise<Array<Collection<any>>>;
+
+	getName(): string;
+
+	// tslint:disable-next-line:no-unnecessary-generics
+	getTable<T = {}>(name: string): Table<T>;
```
Comment (uniqueiniquity) [inline types/mysql__xdevapi/index.d.ts]: Same here, except `Table<{}>`.

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]

**Pull request closed without merging**